### PR TITLE
Wav2vec frontend

### PR DESCRIPTION
**Pull request by Squire-tomsk**

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -314,8 +314,13 @@ def forward(self, x : shaping.BCT, xlen : shaping.B = None, y : shaping.BY = Non
 
 		return self.dict(logits = logits, log_probs = log_probs, olen = olen, **aux)
 
-	def freeze(self, backbone = 0, decoder0 = False):
-		for m in (list(self.backbone[:backbone]) if backbone else []) + (list(self.decoder)[:1] if decoder0 else []):
+	def freeze(self, backbone = 0, decoder0 = False, frontend = False):
+		freezed_modules_list = []
```
Comment (vadimkantorov) [inline models.py]: nit: freezed -> frozen

File: models.py [hunk 2]
```diff
@@ -456,6 +461,47 @@ def sample_rate(self):
...
+			raise RuntimeError(f'Wrong wav2vec aggregator {fairseq_args.aggregator}. Use cnn or gru instead.')
+
+		super().__init__()
+		self.fairseq_args = fairseq_args
+		self.preemphasis = preemphasis
+		self.use_context_features = use_context_features
+		self.model = Wav2VecModel.build_model(fairseq_args, None).eval()
+
+	def forward(self, signal : shaping.BT, mask : shaping.BT = None) -> shaping.BCT:
+		signal = signal if signal.is_floating_point() else signal.to(torch.float32)
```
Comment (vadimkantorov) [inline models.py]: is this needed? i don't remember myself

File: train.py [hunk 3]
```diff
@@ -204,8 +204,17 @@ def main(args):
 			for k in checkpoint['model_state_dict']
 		}
 
+	if args.frontend_checkpoint:
+		frontend_checkpoint = torch.load(args.frontend_checkpoint, map_location = 'cpu')
+		frontend_fariseq_args = frontend_checkpoint['args']
```
Comment (vadimkantorov) [inline train.py]: typo: fariseq -> fairseq. or even better: frontend_extra_args

File: train.py [hunk 4]
```diff
@@ -242,15 +251,16 @@ def main(args):
 		datasets.Labels(lang, bpe = bpe, name = f'bpe{i}', normalize_text_config = normalize_text_config) for i,
 		bpe in enumerate(args.bpe)
 	]
-	frontend = models.LogFilterBankFrontend(
+	frontend = getattr(models, args.frontend)(
 		args.num_input_features,
 		args.sample_rate,
```
Comment (vadimkantorov) [inline train.py]: Please turn all arguments into kwargs. It would be a bit easier to ignore them.

File: train.py [hunk 5]
```diff
@@ -242,15 +251,16 @@ def main(args):
...
 		args.window_size,
 		args.window_stride,
 		args.window,
 		dither = args.dither,
 		dither0 = args.dither0,
-		stft_mode = 'conv' if args.onnx else None
+		stft_mode = 'conv' if args.onnx else None,
+		fairseq_args = frontend_fariseq_args
```
Comment (vadimkantorov) [inline train.py]: You can check if simply passing `**frontend_extra_args` would work (this would depend on whether fairseq conflicts with other kwargs, e.g. window/dither/dither0).
Alternatively you can put all these args into `frontend_extra_args = dict(dither = ..., dither0 = ...)` if LogFilterBank is used and load the args from the frontend checkpoint if not. That should work

File: models.py (hunk 2 above)
Comment (Squire-tomsk) [reply models.py]: Made by analogy with LogFilterBankFrontend. I dont know.

Comment (vadimkantorov) [reply models.py]: This was needed there:
```python
signal = signal if signal.is_floating_point() else signal.to(torch.float32)
signal = normalize_signal(signal) if self.normalize_signal else signal
```
because signal was potentially unnormalized (and in int16 format). If you do not support int16 signals and assume float32 inputs and don't do time-wise normalization inside the frontend, it's better to put at assert and remove the cast (the cast could hide a data loading mistake)
